modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java

```diff
@@ -144,12 +143,41 @@ public void onResponse(ReindexTaskState taskState) {
             @Override
             public void onFailure(Exception e) {
                 if (e instanceof VersionConflictEngineException) {
-                    // TODO: Need to ensure that the allocation has changed
-                    if (isDone.compareAndSet(false, true)) {
-                        onCancel.run();
-                    }
+                    logger.debug(new ParameterizedMessage("Failed to write to {} index on CHECKPOINT due to version conflict, " +
+                        "verifying allocation now", REINDEX_INDEX), e);
+                    reindexIndexClient.getReindexTaskDoc(persistentTaskId, new ActionListener<>() {
+                        @Override
+                        public void onResponse(ReindexTaskState reindexTaskState) {
+                            lastState = reindexTaskState;
```

Comment:
I think we are not allowed to set `lastState` until we find out if another allocation has taken over, since otherwise we risk finishing and the finish write succeeding since it uses the primaryTerm and seqNo from the other allocation.

Also, when updating it from checkpoint, I think the field need to be volatile (since checkpoint writes happen on the scheduler thread).